pkg/ciliumenvoyconfig/exp_controller.go

```diff
@@ -0,0 +1,581 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/cilium/cilium/pkg/envoy"
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/labels"
+	"github.com/cilium/cilium/pkg/loadbalancer"
+	"github.com/cilium/cilium/pkg/loadbalancer/experimental"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/policy/api"
+	"github.com/cilium/cilium/pkg/rate"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+var (
+	experimentalCell = cell.Module(
+		"experimental",
+		"Integration to experimental LB control-plane",
+
+		cell.ProvidePrivate(
+			cecListerWatchers,
+			newPolicyTrigger,
+			func(xds envoy.XDSServer) resourceMutator { return xds },
+		),
+
+		experimentalTableCells,
+		experimentalControllerCells,
+	)
+
+	experimentalControllerCells = cell.Group(
+		cell.Provide(
+			newCECController,
+		),
+		cell.Invoke((*cecController).setWriter),
+		cell.Invoke(registerEnvoyReconciler),
+	)
+
+	experimentalTableCells = cell.Group(
+		cell.ProvidePrivate(
+			NewCECTable,
+			statedb.RWTable[*CEC].ToTable,
+			newNodeLabels,
+		),
+		cell.Invoke(
+			registerCECReflector,
+		),
+	)
+)
+
+type cecControllerParams struct {
+	cell.In
+
+	DB             *statedb.DB
+	JobGroup       job.Group
+	Log            *slog.Logger
+	ExpConfig      experimental.Config
+	LocalNodeStore *node.LocalNodeStore
+
+	NodeLabels    *nodeLabels
+	CECs          statedb.RWTable[*CEC]
+	EnvoySyncer   resourceMutator
+	PolicyTrigger policyTrigger
+}
+
+type resourceMutator interface {
+	DeleteEnvoyResources(context.Context, envoy.Resources) error
+	UpdateEnvoyResources(context.Context, envoy.Resources, envoy.Resources) error
+}
+
+type policyTrigger interface {
+	TriggerPolicyUpdates()
+}
+
+type cecController struct {
+	params cecControllerParams
+	writer *experimental.Writer
+}
+
+type cecControllerOut struct {
+	cell.Out
+
+	C    *cecController
+	Hook experimental.ServiceHook `group:"service-hooks"`
+}
+
+func newNodeLabels() *nodeLabels {
+	nl := &nodeLabels{}
+	lbls := map[string]string{}
+	nl.Store(&lbls)
+	return nl
+}
+
+func newCECController(params cecControllerParams) cecControllerOut {
+	if !params.ExpConfig.EnableExperimentalLB {
+		return cecControllerOut{}
+	}
+
+	c := &cecController{params, nil}
+	params.JobGroup.Add(job.OneShot("proxy-redirect-controller", c.proxyRedirectController))
+	params.JobGroup.Add(job.OneShot("backends-controller", c.backendController))
+	params.JobGroup.Add(job.OneShot("node-label-controller", c.nodeLabelController))
+	return cecControllerOut{
+		C:    c,
+		Hook: c.onServiceUpsert,
+	}
+}
+
+func (c *cecController) setWriter(w *experimental.Writer) {
+	if c != nil {
+		c.writer = w
+	}
+}
+
+// proxyRedirectController watches for changed CECs and updates the proxy redirect in services.
+func (c *cecController) proxyRedirectController(ctx context.Context, health cell.Health) error {
+	wtxn := c.params.DB.WriteTxn(c.params.CECs)
```

Comment:
This dance is awkward to me as an outsider; we take a transaction to get an extra-transaction iterator? And then we also need a new transaction for the iterator to provide us a changestream?

I imagine there are good reasons for, this, but it definitely feels easy to get wrong.

Reply:
There's been a lot of discussion on this. The `Changes` actually modifies the database by registering a "delete tracker" to keep deleted objects to the side so they can be observed. An initial version of this hid the "WriteTxn" call inside "Changes", but I felt that it had potential for deadlocks, so went for the most straightforward and transparent version. https://github.com/cilium/statedb/pull/11#pullrequestreview-2000234227.